jni/src/Foreign/JNI.hs

```diff
@@ -223,6 +228,56 @@ throwIfNull m = do
     then throwIO ArrayCopyFailed
     else return ptr
 
+-- | A system lock allows concurrent readers until the system is shutdown, in
+-- which case no more reading is allowed.
+newtype SystemLock =
+    SystemLock (IORef (Int, SystemDesiredState))
+    -- ^ A count of the actual readers and the desired state of the system.
```

Comment:
Well I don't know what a "system" is. `JvmInstanceLock`? Anyways... it's been merged!